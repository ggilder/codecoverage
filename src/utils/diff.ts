interface FileDiff {
  filename: string
  addedLines: number[]
  deletedLines: number[]
}

export function parseGitDiff(diffOutput: string): FileDiff[] {
  const fileDiffs: FileDiff[] = []
  const lines = diffOutput.split('\n')

  let currentFileDiff: FileDiff | undefined
  let currentAddedLines: number[] = []
  let currentDeletedLines: number[] = []
  let seenHeaderLine = false
  let deletionCurrentLineNumber = 0
  let additionCurrentLineNumber = 0

  for (const line of lines) {
    if (line.startsWith('diff --git')) {
      // New file diff starts
      if (currentFileDiff) {
        currentFileDiff.addedLines = currentAddedLines
        currentFileDiff.deletedLines = currentDeletedLines
        fileDiffs.push(currentFileDiff)
      }

      currentFileDiff = {
        filename: getFilenameFromDiffHeader(line),
        addedLines: [],
        deletedLines: []
      }
      currentAddedLines = []
      currentDeletedLines = []
      seenHeaderLine = false
    } else if (line.startsWith('@@')) {
      // Header line
      seenHeaderLine = true
      const lineInfo = getLineInfoFromHeaderLine(line)
      deletionCurrentLineNumber = lineInfo.deletionStartingLineNumber
      additionCurrentLineNumber = lineInfo.additionStartingLineNumber
    } else if (line.startsWith('+') && seenHeaderLine) {
      // Added line
      currentAddedLines.push(additionCurrentLineNumber)
      additionCurrentLineNumber++
    } else if (line.startsWith('-') && seenHeaderLine) {
      // Deleted line
      currentDeletedLines.push(deletionCurrentLineNumber)
      deletionCurrentLineNumber++
    } else if (seenHeaderLine) {
      // Context line
      deletionCurrentLineNumber++
      additionCurrentLineNumber++
    }
  }

  // Add the last file diff
  if (currentFileDiff) {
    currentFileDiff.addedLines = currentAddedLines
    currentFileDiff.deletedLines = currentDeletedLines
    fileDiffs.push(currentFileDiff)
  }

  return fileDiffs
}

function getFilenameFromDiffHeader(header: string): string {
  // Extract the filename from the diff header
  const startIndex = header.indexOf(' a/') + 3
  const endIndex = header.indexOf(' b/', startIndex)
  const filename = header.substring(startIndex, endIndex)
  return filename
}

function getLineInfoFromHeaderLine(line: string): {
  deletionStartingLineNumber: number
  additionStartingLineNumber: number
} {
  // Extract the starting line numbers for each side of the diff
  const matches = line.match(/-(\d+),?(\d+)? \+(\d+),?(\d+)? @@/)
  if (matches && matches.length === 5) {
    const deletionStartingLineNumber = parseInt(matches[1], 10)
    const additionStartingLineNumber = parseInt(matches[3], 10)
    return {deletionStartingLineNumber, additionStartingLineNumber}
  }
  return {deletionStartingLineNumber: 0, additionStartingLineNumber: 0}
}
