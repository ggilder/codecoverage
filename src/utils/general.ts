export function filterCoverageByFile(coverage: CoverageParsed): CoverageFile[] {
  return coverage.map(item => ({
    fileName: item.file,
    missingLineNumbers: item?.lines?.details
      .filter(line => line.hit === 0)
      .map(line => line.line)
  }))
}

export function coalesceLineNumbers(lineNumbers: number[]): LineRange[] {
  const ranges: LineRange[] = []
  let rstart
  let rend
  for (let i = 0; i < lineNumbers.length; i++) {
    rstart = lineNumbers[i]
    rend = rstart
    while (lineNumbers[i + 1] - lineNumbers[i] === 1) {
      rend = lineNumbers[i + 1]
      i++
    }
    ranges.push({start_line: rstart, end_line: rend})
  }
  return ranges
}

export function intersectLineRanges(
  a: LineRange[],
  b: LineRange[]
): LineRange[] {
  const result: LineRange[] = []
  let i = 0
  let j = 0

  while (i < a.length && j < b.length) {
    const rangeA = a[i]
    const rangeB = b[j]

    if (rangeA.end_line < rangeB.start_line) {
      i++
    } else if (rangeB.end_line < rangeA.start_line) {
      j++
    } else {
      const start = Math.max(rangeA.start_line, rangeB.start_line)
      const end = Math.min(rangeA.end_line, rangeB.end_line)
      result.push({start_line: start, end_line: end})

      if (rangeA.end_line < rangeB.end_line) {
        i++
      } else {
        j++
      }
    }
  }

  return result
}

export function correctLineTotals(coverage: CoverageParsed): CoverageParsed {
  return coverage.map(item => ({
    ...item,
    lines: {
      ...item.lines,
      found: item.lines.details.length,
      hit: item.lines.details.filter(line => line.hit > 0).length
    }
  }))
}

export type CoverageParsed = {
  file: string
  title: string
  lines: {
    found: number
    hit: number
    details: {
      line: number
      hit: number
      name: string
    }[]
  }
}[]

export type CoverageFile = {
  fileName: string
  missingLineNumbers: number[]
}

export type LineRange = {
  start_line: number
  end_line: number
}
