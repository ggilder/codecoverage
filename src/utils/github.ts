import * as core from '@actions/core'
import * as diff from './diff'
import * as github from '@actions/github'
import {
  CoverageFile,
  LineRange,
  coalesceLineNumbers,
  intersectLineRanges
} from './general'
import {Octokit} from 'octokit'

export class GithubUtil {
  private client: Octokit

  constructor(token: string) {
    if (!token) {
      throw new Error('GITHUB_TOKEN is missing')
    }
    this.client = new Octokit({auth: token})
  }

  getPullRequestRef(): string {
    const pullRequest = github.context.payload.pull_request
    return pullRequest
      ? pullRequest.head.ref
      : github.context.ref.replace('refs/heads/', '')
  }

  async getPullRequestDiff(): Promise<PullRequestFiles> {
    const pull_number = github.context.issue.number
    const response = await this.client.rest.pulls.get({
      ...github.context.repo,
      pull_number,
      mediaType: {
        format: 'diff'
      }
    })
    // @ts-expect-error With mediaType param, response.data is actually a string, but the response type doesn't reflect this
    const fileLines = diff.parseGitDiff(response.data)
    const prFiles: PullRequestFiles = {}
    for (const item of fileLines) {
      prFiles[item.filename] = coalesceLineNumbers(item.addedLines)
    }

    return prFiles
  }

  /**
   * https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-warning-message
   */
  annotate(annotations: Annotations[]): void {
    for (const ann of annotations) {
      console.log(
        `::${ann.annotation_level} file=${ann.path},line=${ann.start_line},endLine=${ann.end_line}::${ann.message}`
      )
    }
  }

  buildAnnotations(
    coverageFiles: CoverageFile[],
    pullRequestFiles: PullRequestFiles
  ): Annotations[] {
    const annotations: Annotations[] = []
    for (const current of coverageFiles) {
      // Only annotate relevant files
      const prFileRanges = pullRequestFiles[current.fileName]
      if (prFileRanges) {
        const coverageRanges = coalesceLineNumbers(current.missingLineNumbers)
        const uncoveredRanges = intersectLineRanges(
          coverageRanges,
          prFileRanges
        )

        // Only annotate relevant line ranges
        for (const uRange of uncoveredRanges) {
          const message =
            uRange.end_line > uRange.start_line
              ? 'These lines are not covered by a test'
              : 'This line is not covered by a test'
          annotations.push({
            path: current.fileName,
            start_line: uRange.start_line,
            end_line: uRange.end_line,
            annotation_level: 'warning',
            message
          })
        }
      }
    }
    core.info(`Annotation count: ${annotations.length}`)
    return annotations
  }
}

type Annotations = {
  path: string
  start_line: number
  end_line: number
  start_column?: number
  end_column?: number
  annotation_level: string
  message: string
}

type PullRequestFiles = {
  [key: string]: LineRange[]
}
