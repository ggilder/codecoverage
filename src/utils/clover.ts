import * as clover from '@cvrg-report/clover-json'
import * as fs from 'fs'
import * as path from 'path'
import {CoverageParsed} from './general'

export async function parseClover(
  coveragePath: string,
  workspacePath: string
): Promise<CoverageParsed> {
  if (!coveragePath) {
    throw Error('No Clover XML path provided')
  }
  if (!workspacePath) {
    throw Error('No workspace path provided')
  }
  const fileRaw = fs.readFileSync(coveragePath, 'utf8')
  const parsed = await clover.parseContent(fileRaw)

  for (const entry of parsed) {
    entry.file = path.relative(workspacePath, entry.path || entry.file)
  }

  return parsed
}
