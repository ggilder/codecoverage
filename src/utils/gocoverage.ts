import * as fs from 'fs'
import * as gocov from 'golang-cover-parse'
import * as path from 'path'
import * as readline from 'readline'
import {CoverageParsed} from './general'

export async function parseGoCoverage(
  coveragePath: string,
  goModPath: string
): Promise<CoverageParsed> {
  if (!coveragePath) {
    throw Error('No Go coverage path provided')
  }

  if (!goModPath) {
    throw Error('No Go module path provided')
  }

  const goModule = await parseGoModFile(goModPath)

  const fileRaw = fs.readFileSync(coveragePath, 'utf8')
  return new Promise((resolve, reject) => {
    gocov.parseContent(fileRaw, (err, result) => {
      if (err === null) {
        filterModulePaths(result, goModule)
        resolve(result)
      }
      reject(err)
    })
  })
}

function filterModulePaths(entries, moduleName: string): void {
  for (const entry of entries) {
    entry.file = path.relative(moduleName, entry.file)
  }
}

async function parseGoModFile(filePath: string): Promise<string> {
  const fileStream = fs.createReadStream(filePath)
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  })
  // Note: we use the crlfDelay option to recognize all instances of CR LF
  // ('\r\n') in input.txt as a single line break.

  for await (const line of rl) {
    if (line.startsWith('module ')) {
      return line.slice(7)
    }
  }

  /* istanbul ignore next */
  return ''
}
