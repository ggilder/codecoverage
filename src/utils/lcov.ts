import * as NodeUtil from 'util'
import * as fs from 'fs'
import * as path from 'path'
import {CoverageParsed} from './general'
import parser from 'lcov-parse'

export async function parseLCov(
  lcovPath: string,
  workspacePath: string
): Promise<CoverageParsed> {
  if (!lcovPath) {
    throw Error('No LCov path provided')
  }

  const parserPromise = NodeUtil.promisify(parser)
  const fileRaw = fs.readFileSync(lcovPath, 'utf8')
  const parsed = (await parserPromise(fileRaw)) as CoverageParsed

  for (const entry of parsed) {
    entry.file = path.relative(workspacePath, entry.file)
  }

  return parsed
}
