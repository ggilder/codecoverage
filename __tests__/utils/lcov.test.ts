import {test} from '@jest/globals'
import {parseLCov} from '../../src/utils/lcov'
import {getFixturePath} from '../fixtures/util'

test('should parse lCov file', async function () {
  const path = getFixturePath('lcov.info')
  const output = await parseLCov(path, process.cwd())
  expect(output).toMatchSnapshot()
})

test('should throw err if lCov file path is not given', async function () {
  await expect(parseLCov('', '')).rejects.toThrow('No LCov path provided')
})
