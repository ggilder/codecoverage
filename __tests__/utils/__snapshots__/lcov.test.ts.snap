// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should parse lCov file 1`] = `
[
  {
    "branches": {
      "details": [
        {
          "block": 0,
          "branch": 0,
          "line": 3,
          "taken": 0,
        },
        {
          "block": 0,
          "branch": 1,
          "line": 3,
          "taken": 0,
        },
        {
          "block": 1,
          "branch": 0,
          "line": 3,
          "taken": 0,
        },
        {
          "block": 1,
          "branch": 1,
          "line": 3,
          "taken": 0,
        },
        {
          "block": 2,
          "branch": 0,
          "line": 3,
          "taken": 0,
        },
        {
          "block": 2,
          "branch": 1,
          "line": 3,
          "taken": 0,
        },
        {
          "block": 3,
          "branch": 0,
          "line": 3,
          "taken": 0,
        },
        {
          "block": 3,
          "branch": 1,
          "line": 3,
          "taken": 0,
        },
        {
          "block": 4,
          "branch": 0,
          "line": 4,
          "taken": 0,
        },
        {
          "block": 4,
          "branch": 1,
          "line": 4,
          "taken": 0,
        },
        {
          "block": 5,
          "branch": 0,
          "line": 4,
          "taken": 0,
        },
        {
          "block": 5,
          "branch": 1,
          "line": 4,
          "taken": 0,
        },
        {
          "block": 6,
          "branch": 0,
          "line": 4,
          "taken": 0,
        },
        {
          "block": 6,
          "branch": 1,
          "line": 4,
          "taken": 0,
        },
        {
          "block": 7,
          "branch": 0,
          "line": 4,
          "taken": 0,
        },
        {
          "block": 7,
          "branch": 1,
          "line": 4,
          "taken": 0,
        },
      ],
      "found": 16,
      "hit": 0,
    },
    "file": "src/utils/general.ts",
    "functions": {
      "details": [
        {
          "hit": 0,
          "line": 2,
          "name": "getFileNameFirstItemFromPath",
        },
      ],
      "found": 1,
      "hit": 0,
    },
    "lines": {
      "details": [
        {
          "hit": 1,
          "line": 2,
        },
        {
          "hit": 0,
          "line": 3,
        },
        {
          "hit": 0,
          "line": 4,
        },
      ],
      "found": 3,
      "hit": 1,
    },
    "title": "",
  },
  {
    "branches": {
      "details": [
        {
          "block": 0,
          "branch": 0,
          "line": 11,
          "taken": 0,
        },
        {
          "block": 1,
          "branch": 0,
          "line": 19,
          "taken": 0,
        },
        {
          "block": 1,
          "branch": 1,
          "line": 19,
          "taken": 0,
        },
        {
          "block": 2,
          "branch": 0,
          "line": 36,
          "taken": 0,
        },
        {
          "block": 2,
          "branch": 1,
          "line": 36,
          "taken": 0,
        },
        {
          "block": 3,
          "branch": 0,
          "line": 36,
          "taken": 0,
        },
        {
          "block": 3,
          "branch": 1,
          "line": 36,
          "taken": 0,
        },
        {
          "block": 4,
          "branch": 0,
          "line": 37,
          "taken": 0,
        },
        {
          "block": 5,
          "branch": 0,
          "line": 67,
          "taken": 0,
        },
        {
          "block": 5,
          "branch": 1,
          "line": 67,
          "taken": 0,
        },
        {
          "block": 6,
          "branch": 0,
          "line": 67,
          "taken": 0,
        },
        {
          "block": 6,
          "branch": 1,
          "line": 67,
          "taken": 0,
        },
        {
          "block": 7,
          "branch": 0,
          "line": 68,
          "taken": 0,
        },
        {
          "block": 8,
          "branch": 0,
          "line": 68,
          "taken": 0,
        },
        {
          "block": 8,
          "branch": 1,
          "line": 68,
          "taken": 0,
        },
      ],
      "found": 15,
      "hit": 0,
    },
    "file": "src/utils/github.ts",
    "functions": {
      "details": [
        {
          "hit": 1,
          "line": 10,
          "name": "(anonymous_13)",
        },
        {
          "hit": 0,
          "line": 17,
          "name": "(anonymous_14)",
        },
        {
          "hit": 0,
          "line": 27,
          "name": "(anonymous_15)",
        },
        {
          "hit": 0,
          "line": 42,
          "name": "(anonymous_17)",
        },
        {
          "hit": 0,
          "line": 60,
          "name": "(anonymous_19)",
        },
        {
          "hit": 0,
          "line": 69,
          "name": "(anonymous_20)",
        },
      ],
      "found": 6,
      "hit": 1,
    },
    "lines": {
      "details": [
        {
          "hit": 1,
          "line": 1,
        },
        {
          "hit": 1,
          "line": 2,
        },
        {
          "hit": 1,
          "line": 4,
        },
        {
          "hit": 1,
          "line": 5,
        },
        {
          "hit": 1,
          "line": 7,
        },
        {
          "hit": 1,
          "line": 11,
        },
        {
          "hit": 0,
          "line": 12,
        },
        {
          "hit": 1,
          "line": 14,
        },
        {
          "hit": 0,
          "line": 18,
        },
        {
          "hit": 0,
          "line": 19,
        },
        {
          "hit": 0,
          "line": 28,
        },
        {
          "hit": 0,
          "line": 29,
        },
        {
          "hit": 0,
          "line": 33,
        },
        {
          "hit": 0,
          "line": 34,
        },
        {
          "hit": 0,
          "line": 35,
        },
        {
          "hit": 0,
          "line": 36,
        },
        {
          "hit": 0,
          "line": 37,
        },
        {
          "hit": 0,
          "line": 39,
        },
        {
          "hit": 0,
          "line": 40,
        },
        {
          "hit": 0,
          "line": 44,
        },
        {
          "hit": 0,
          "line": 56,
        },
        {
          "hit": 0,
          "line": 57,
        },
        {
          "hit": 0,
          "line": 64,
        },
        {
          "hit": 0,
          "line": 65,
        },
        {
          "hit": 0,
          "line": 67,
        },
        {
          "hit": 0,
          "line": 68,
        },
        {
          "hit": 0,
          "line": 69,
        },
        {
          "hit": 0,
          "line": 70,
        },
        {
          "hit": 0,
          "line": 82,
        },
        {
          "hit": 0,
          "line": 83,
        },
      ],
      "found": 30,
      "hit": 7,
    },
    "title": "",
  },
  {
    "branches": {
      "details": [
        {
          "block": 0,
          "branch": 0,
          "line": 6,
          "taken": 0,
        },
        {
          "block": 1,
          "branch": 0,
          "line": 17,
          "taken": 0,
        },
        {
          "block": 1,
          "branch": 1,
          "line": 17,
          "taken": 4,
        },
        {
          "block": 2,
          "branch": 0,
          "line": 17,
          "taken": 4,
        },
        {
          "block": 2,
          "branch": 1,
          "line": 17,
          "taken": 4,
        },
        {
          "block": 3,
          "branch": 0,
          "line": 17,
          "taken": 0,
        },
        {
          "block": 3,
          "branch": 1,
          "line": 17,
          "taken": 4,
        },
        {
          "block": 4,
          "branch": 0,
          "line": 17,
          "taken": 4,
        },
        {
          "block": 4,
          "branch": 1,
          "line": 17,
          "taken": 4,
        },
      ],
      "found": 9,
      "hit": 6,
    },
    "file": "src/utils/lcov.ts",
    "functions": {
      "details": [
        {
          "hit": 0,
          "line": 5,
          "name": "parseLCov",
        },
        {
          "hit": 1,
          "line": 14,
          "name": "filterCoverageByFile",
        },
        {
          "hit": 4,
          "line": 15,
          "name": "(anonymous_17)",
        },
        {
          "hit": 18,
          "line": 18,
          "name": "(anonymous_18)",
        },
        {
          "hit": 12,
          "line": 19,
          "name": "(anonymous_19)",
        },
      ],
      "found": 5,
      "hit": 4,
    },
    "lines": {
      "details": [
        {
          "hit": 1,
          "line": 1,
        },
        {
          "hit": 1,
          "line": 2,
        },
        {
          "hit": 1,
          "line": 3,
        },
        {
          "hit": 1,
          "line": 5,
        },
        {
          "hit": 0,
          "line": 6,
        },
        {
          "hit": 0,
          "line": 7,
        },
        {
          "hit": 0,
          "line": 9,
        },
        {
          "hit": 0,
          "line": 10,
        },
        {
          "hit": 0,
          "line": 11,
        },
        {
          "hit": 1,
          "line": 14,
        },
        {
          "hit": 4,
          "line": 15,
        },
        {
          "hit": 18,
          "line": 18,
        },
        {
          "hit": 12,
          "line": 19,
        },
      ],
      "found": 13,
      "hit": 8,
    },
    "title": "",
  },
]
`;
